### ISSUE TITLE
Split helper functions for input validation and reshaping

### ISSUE THREAD

**Issue by kachayev** [scikit-adaptation/skada#58]
As of now we have the following input validation helpers:

* `check_X_y_domain`
* `check_X_domain`

They are designed to be used similarly to sklearn functions like `utils.check_X_y` ([link](https://scikit-learn.org/stable/modules/generated/sklearn.utils.check_X_y.html)) with an advantage of understanding `sample_domain` parameter (or inferring it when necessary).

Right now they provide multiple different output shapes, like the option to split sources vs. target or return only domain indices. What need to be done is the following:

* `check_X_y_domain` always returns a tuple of `X, y, sample_domain`, `check_X_domain`, respectively, always returns `X, sample_domain`
* both are moved to a publicly available namespace `utils` instead of internal `_utils` (to match sklearn counterparts)
* separately, new utils functions are introduced to cover up previously available functionality (split sources vs. targets, extracting indices, etc)

Don't forget about proper docstrings and tests coverage.

**Comment by kachayev** [comment]
Implementation in #53
